src/deluge/vm/wrenimpl.cpp

```diff
@@ -0,0 +1,238 @@
+#if ENABLE_WREN
+
+#include "hid/display/numeric_driver.h"
+#include "storage/storage_manager.h"
+#include "memory/general_memory_allocator.h"
+#include "wrenimpl.h"
+#include "api.h"
+#include "heap.h"
+#include <cstring>
+
+extern "C" {
+#include "drivers/uart/uart.h"
+}
+
+namespace Wren {
+
+static char scriptBuffer[SCRIPT_BUFFER_SIZE];
+
+void VM::print(const char* text) {
+	bool empty = true;
+	for (size_t i = 0; text[i] != '\0'; i++) {
+		if (!std::isspace(static_cast<unsigned char>(text[i]))) {
+			empty = false;
+		}
+	}
+	if (empty) {
+		return;
+	}
+
+#if HAVE_OLED
+	numericDriver.displayPopup(text);
+#else
+	if (strlen(text) <= NUMERIC_DISPLAY_LENGTH) {
+		numericDriver.setText(text, true);
+	}
+	else {
+		numericDriver.setScrollingText(text);
+	}
+#endif
+}
+
+void VM::writeFn(WrenVM* vm, const char* text) {
+	VM::print(text);
+}
+
+//#define WREN_LOG(x) len = strlen(x); if(len > 200) len = 200; f_write(&fil, x, len, &written);
+
+void VM::errorFn(WrenVM* vm, WrenErrorType errorType, const char* mod, const int line, const char* msg) {
+	//UINT written;
+	//FIL fil;
+	//auto res = storageManager.createFile(&fil, "SCRIPTS/log.txt", true);
+	//size_t len;
+
+	switch (errorType) {
+	case WREN_ERROR_COMPILE:
+		VM::print("E compile");
+
+		uartPrint(mod);
```

Comment:
it would be useful to switch to the new highlevel `Debug::print()` family of functions to allow debugging wren over USB.